src/librustc_trans/back/msvc/mod.rs

```diff
@@ -223,31 +256,25 @@ mod platform {
     // where someone on 32-bit Windows is trying to cross compile to 64-bit and
     // it tries to invoke the native 64-bit linker which won't work.
     //
-    // FIXME - This currently functions based on the host architecture of rustc
-    // itself but it should instead detect the bitness of the OS itself.
+    // When possible we prefer the linker whose host is 64-bit, even if Rust
+    // itself is 32-bit and the target is 32-bit, since that gives the linker
+    // more address space to work with and is probably faster.
     //
     // FIXME - Figure out what happens when the host architecture is arm.
-    //
-    // FIXME - Some versions of MSVC may not come with all these toolchains.
-    // Consider returning an array of toolchains and trying them one at a time
-    // until the linker is found.
-    fn bin_subdir(arch: &str) -> Option<&'static str> {
-        if cfg!(target_arch = "x86_64") {
-            match arch {
-                "x86" => Some("amd64_x86"),
-                "x86_64" => Some("amd64"),
-                "arm" => Some("amd64_arm"),
-                _ => None,
-            }
-        } else if cfg!(target_arch = "x86") {
-            match arch {
-                "x86" => Some(""),
-                "x86_64" => Some("x86_amd64"),
-                "arm" => Some("x86_arm"),
-                _ => None,
+    fn bin_subdir(arch: &str) -> Vec<(&'static str, &'static str)> {
+        if let Some(host) = host_arch() {
+            match (arch, host) {
+                ("x86", Arch::X86) => vec![("", "")],
+                ("x86", Arch::Amd64) => vec![("amd64_x86", "amd64"), ("", "")],
+                ("x86_64", Arch::X86) => vec![("x86_amd64", "")],
+                ("x86_64", Arch::Amd64) => vec![("amd64", "amd64"), ("x86_amd64", "")],
+                ("arm", Arch::X86) => vec![("x86_arm", "")],
+                ("arm", Arch::Amd64) => vec![("amd64_arm", "amd64"), ("x86_arm", "")],
+                _ => vec![],
             }
-        } else { None }
+        } else { vec![] }
```

Comment:
Perhaps this could just be:

```rust
match (arch, host_arch()) {
    // ... 
}
```

Can you also add docs as to what's being returned here? It's not clear to me at least what these are vectors of nor what each element of the pair is.